src/main/scala/org/constellation/primitives/ConcurrentTipService.scala

```diff
@@ -213,4 +215,43 @@ class ConcurrentTipService[F[_]: Concurrent: Logger: Clock](
       }
     selectedFacils.toMap
   }
+
+  private def filterFacilitators(facilitators: Map[Id, PeerData]) =
+    for {
+      facilitatorHeights <- LiftIO[F].liftIO(getFacilitatorHeights(facilitators))
+      ownHeight <- ownHeight()
+      _ <- Logger[F].debug(s"filter facilitators : own height = $ownHeight : facilitatorHeights = $facilitatorHeights")
+
+      facilitatorsIds <- filterByHeight(facilitatorHeights, maxOrZero(ownHeight))
+      _ <- Logger[F].debug(s"facilitators ids = $facilitatorsIds")
+
+      filteredFacilitators <- Sync[F].delay { facilitators.filter(f => facilitatorsIds.contains(f._1)) }
+      _ <- Logger[F].info(s"filtered facilitators = $filteredFacilitators")
+    } yield filteredFacilitators
+
+  private def getFacilitatorHeights(facilitators: Map[Id, PeerData]) =
+    facilitators.toList
+      .traverse(
+        p =>
+          (
+            p._1,
+            p._2.client
+              .getNonBlockingIO[List[Height]]("heights")(IO.contextShift(ConstellationExecutionContext.bounded))
+          ).sequence
+      )
+
+  private def ownHeight() =
+    for {
+      tips <- toMap
+      maybeHeights <- LiftIO[F].liftIO(tips.toList.traverse(t => dao.checkpointService.lookup(t._1)))
```

Comment:
Can't `checkpointService` be passed in parameters?

Reply:
i can't do it because `checkpointService` contains `ConcurrentTipService`, cyclic dependency

Reply:
no worries since I have extracted CheckpointService -> CheckpointService, CheckpointAcceptanceService tips service is only required in acceptance service